### PR TITLE
Fix node-nodebox lighting difference in direct sunlight

### PR DESCRIPTION
**Pull request by numberZero**
Fixes smooth lighting part of #7040

1. In `getSmoothLightCombined` (which is called for each corner of the node), it now checks for direct sunlight (light level of 15). If it is detected, full brightness (255) is used instead of the average. Then, ambient occlusion is applied, if necessary; otherwise it returns 255 for the day light bank.
2. `MapblockMeshGenerator::getSmoothLightFrame` checks for that value and marks the corresponding corner **and** its bottom counterpart (for convenience) as exposed to direct sunlight.
3. `MapblockMeshGenerator::blendLight` now calculates two values for the day light bank: a base light value (used for sides), computed as usual, and a boosted value (used for sun-facing quads), computed considering marked corners as having full light.
4. `LightInfo::getPair` blends these values using given factor, for which dot product of sun direction (0, 1, 0) and face normal, clamped to [0, 1], is used. This way, sides and bottom faces are not affected, but top faces get full sunlight. This feature is only used for nodeboxes and mesh nodes.

Note: lighting settings on screenshot are **extreme** to make defects, if any, more noticeable. That’s the reason for such dark shadow.
screenshot_20180220_160414

### CODE REVIEW

**Comment by Fixer-007** [comment]
Do I need to apply 7051 as well?

**Comment by numberZero** [comment]
#7051 is for flat lighting. This one is for smooth lighting. They
happened to be completely independent.

**Comment by Fixer-007** [comment]
This PR seems good to me, I've checked it in game.

**Comment by paramat** [comment]
From IRC http://irc.minetest.net/minetest-dev/2018-02-20#i_5234396

16:05 numzero   okay, now #7061 works for nodeboxes and meshnodes; other cuboid-using nodes (framed glass) are affected but should be fine as well
16:05 ShadowBot https://github.com/minetest/minetest/issues/7061 -- Fix regression after #6696 by numberZero
16:07 numzero   I saw a glitch on extreme settings (lighting_beta=4.0 and plain white texture) but even in such conditions it is barely noticeable
16:07 numzero   it also may be GPU-specific
16:08 numzero   as it is fixed by moving one multiplication from the vertex shader to the fragment one (with unlimited precision, that move would not change anything)
16:10 numzero   but without shaders such change is not possible ofc
16:12 numzero   note that it doesn’t affect flat lighting, my previous PR fixes that

**Comment by paramat** [comment]
Don't understand the code but will test.

**Comment by paramat** [comment]
Please can you help reviewers by explaining in the first post what this does, how it works and why it works this way? We're short on reviewer time and available reviewers may not be good with this area of code, if i understand how this works i may be able to +1 it.

**Comment by numberZero** [comment]
@paramat Done. I should probably document that in the source briefly, though.

**Comment by paramat** [comment]
Sorry for delay, i will test this as soon as possible.

**Comment by rubenwardy** [comment]
Change set looks good, although I'm not familiar with this area of the code

**Comment by SmallJoker** [comment]
Side-by-side comparison using the default settings: 
screenshot_20180310_095322

Looks much better now.

**Comment by paramat** [comment]
Tested your branch, there is a bug with smooth lighting disabled but i assume that's because #7051 is not included in this branch.

With smooth lighting and a white snow texture, snow slabs next to snowblocks:

screenshot_20180314_013941

**Comment by paramat** [comment]
screenshot_20180314_020252

screenshot_20180314_020403

Ambient occlusion is different for different corners.

**Comment by paramat** [comment]
Maybe i should test master plus your patch instead.

**Comment by paramat** [comment]
Applying patch to master, same appearence, all else looks fine though:

screenshot_20180314_022814

**Comment by paramat** [comment]
In master appearence is the same, so this is not caused by your mod:

screenshot_20180314_023600

So this PR tests fine.

**Comment by numberZero** [comment]
> Ambient occlusion is different for different corners.

@paramat It isn’t. Each quad is drawn as a pair of triangles (check the wireframe view), and the same diagonal is used for that split, regardless of node position; that is, it is aligned identically relative to the world, but therefore differently relative to the corners. That’s not a (new) bug if these are slabs; there was some code to choose the diagonal based on lighting to make gradients look nicer, but it was for fast-faces (solid nodes) only. It might be possible to do the same for nodeboxes, however.

**Comment by paramat** [comment]
Yes i suspected that was nothing new, good to know thanks.
Your PR tests fine then, will look at code again.

**Comment by paramat** [comment]
I'll look at code then will probably +1 and merge it.

File: src/content_mapblock.cpp [hunk 1]
```diff
@@ -280,45 +282,61 @@ void MapblockMeshGenerator::drawCuboid(const aabb3f &box,
...
 		LightPair light(getSmoothLightTransparent(blockpos_nodes + p, light_dirs[k], data));
-		frame.lightsA[k] = light.lightA;
-		frame.lightsB[k] = light.lightB;
+		frame.lightsDay[k] = light.lightDay;
+		frame.lightsNight[k] = light.lightNight;
+		// If there is direct sunlight and no ambient occlusion at some corner,
+		// mark the vertical edge (top and bottom corners) containing it.
+		if (light.lightDay == 255) {
+			frame.sunlight[k] = true;
+			frame.sunlight[k ^ 2] = true;
```
Comment (paramat) [inline src/content_mapblock.cpp]: I'm just wondering how k ^ 2 will get the bottom corner vertex number here from the top corner number?

**Comment by paramat** [comment]
Thanks to your description i have a vague understanding of this, but see line comment.
Once i understand the k ^ 2 thing i will +1.
@numberZero

File: src/content_mapblock.cpp (hunk 1 above)
Comment (paramat) [reply src/content_mapblock.cpp]: Nevermind, '^' is XOR not power.

**Comment by nerzhul** [comment]
Merged, thanks

**Pull request merged**